README.MD
# ngx-qrcode2

`ngx-qrcode2` An Angular 6  Component library for Generating QR (Quick Response ) Codes.

You can use the `ngx-qrcode2` to easily generate QR codes inside your Angular 6 or Ionic 3 applications 

>QR code (abbreviated from Quick Response Code) is the trademark for a type of matrix barcode (or two-dimensional barcode) first designed for the automotive industry in Japan. A barcode is a machine-readable optical label that contains information about the item to which it is attached. A QR code uses four standardized encoding modes (numeric, alphanumeric, byte/binary, and kanji) to efficiently store data; extensions may also be used. [Source](https://en.wikipedia.org/wiki/QR_code)

## How to install ngx-qrcode2?

To use ngx-qrcode in your project, install it via npm:

```bash
$ npm install ngx-qrcode2 --save
```

## How to use ngx-qrcode2?

For a complete and detailed tutorial on how to use this library. See:

[How to Generate QR Codes In Angular 4+ Applications ](https://www.techiediaries.com/generate-qrcodes-angular)


Import `NgxQRCodeModule` from `ngx-qrcode2`  into your `app.module.ts`:

```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { NgxQRCodeModule } from 'ngx-qrcode2';

import { AppComponent } from './app.component';


@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    NgxQRCodeModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```


Once the library is imported, you can use the ngx-qrcode component in your Angular application.

In `app.component.html`, add:
```xml

<div style="text-align:center">
  <h1>
    ngx-qrcode2 demo 
  </h1>
</div>

<ngx-qrcode 
      [qrc-element-type]="elementType" 
      [qrc-value] = "value"
      qrc-class = "aclass"
      qrc-errorCorrectionLevel = "L">
</ngx-qrcode>

```
In `app.component.ts`, add:

```ts
import { Component } from '@angular/core';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  title = 'app';
  elementType = 'url';
  value = 'Techiediaries';
}
```


## How to develop ngx-qrcode2?

To generate all `*.js`, `*.d.ts` and `*.metadata.json` files:

```bash
$ npm run build
```

To lint all `*.ts` files:

```bash
$ npm run lint
```

## License

MIT © [Techiediaries](mailto:techiediaries9@gmail.com)


CHANGELOG.md
# 13.0.0 (2022-05-31)


### Build System

* **release:** update to angular 13 ([8ab439a](https://github.com/techiediaries/ngx-qrcode/commit/8ab439a8ed815c01293855eff86072af10460a7d))


### BREAKING CHANGES

* **lib:** support for angular v13
* Replace tslint with eslint
* Change protractor to cypress
* Update third party libraries
* Update node lts version


# [9.1.0](https://github.com/techiediaries/ngx-qrcode/compare/9.0.0...9.1.0) (2020-08-11)


### Code Refactoring

* **packaging:** rename NPM to @techiediaries/ngx-qrcode ([932d26d](https://github.com/techiediaries/ngx-qrcode/commit/932d26da164d44fe056db11f927198aae54860b7)), closes [#47](https://github.com/techiediaries/ngx-qrcode/issues/47)


### Features

* **changelog:** add changelog generation ([0e962ca](https://github.com/techiediaries/ngx-qrcode/commit/0e962ca3567cbd4560926a43c9fe6739abc71525))


### BREAKING CHANGES

* **packaging:** rename this package, @techiediaries/ngx-qrcode must be used instead



<a name="9.0.0"></a>
# 9.0.0 (2020-04-28)


### Bug Fixes

* **core:** Remove the alt attribute when undefined.

### Features

* **lib:** support for angular v9
* **core:** Create unit tests with 100% of code coverage
* **core:** Fix lint

### Refactor

* **core:** Create types, constants and enums to type component inputs.

### BREAKING CHANGES

* **lib:** Rename properties to follow angular style guide

Before:
```
[qrc-element-type] 
[qrc-class] 
[qrc-alt] 
[qrc-value] 
[qrc-version]
[qrc-errorCorrectionLevel]
[qrc-margin]
[qrc-scale]
[qrc-width]
[qrc-colorDark]
[qrc-colorLight]
```

After:
```
[elementType]
[cssClass]
[alt]
[value]
[version]
[errorCorrectionLevel]
[margin]
[scale]
[width]
[colorDark]
[colorLight]
```


src/ngx-qrcode2.component.ts
import { Component, Input, OnChanges, ViewChild, Renderer2, ElementRef } from '@angular/core';

declare var require: any;

let QRCode = require('qrcode');

@Component({
  selector: 'ngx-qrcode',
  template: `<div #qrcElement [class]="cssClass"></div>`,
  styles: []
})
export class NgxQRCodeComponent implements OnChanges {

  @Input('qrc-element-type') elementType: 'url' | 'img' | 'canvas' = 'url';
  @Input('qrc-class') cssClass = 'qrcode';
  @Input('qrc-alt') alt: '';
  @Input('qrc-value') value = 'https://www.techiediaries.com';
  @Input('qrc-version') version: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12' | '13' | '14' | '15' | '16' | '17' | '18' | '19' | '20' | '21' | '22' | '23' | '24' | '25' | '26' | '27' | '28' | '29' | '30' | '31' | '32' | '33' | '34' | '35' | '36' | '37' | '38' | '39' | '40' | '' = '';

  @Input('qrc-errorCorrectionLevel') errorCorrectionLevel: 'L' | 'M' | 'Q' | 'H' = 'M';
  @Input('qrc-margin') margin = 4;
  @Input('qrc-scale') scale = 4;
  @Input('qrc-width') width = 10;
  @Input('qrc-colorDark') colorDark = '#000000';
  @Input('qrc-colorLight') colorLight = '#ffffff';

  @ViewChild('qrcElement') qrcElement: ElementRef;


  constructor(private renderer: Renderer2) {

  }

  ngOnChanges() {
    this.createQRCode();
  }
  toDataURL() {
    return new Promise((resolve, reject) => {
      QRCode.toDataURL(this.value,
        {
          version: this.version,
          errorCorrectionLevel: this.errorCorrectionLevel,
          margin: this.margin,
          scale: this.scale,
          width: this.width,
          color: {
            dark: this.colorDark,
            light: this.colorLight
          }
        }, function (err, url) {
          if (err) {
            console.error(err);
            reject(err);
          } else {
            //console.log(url);
            resolve(url);
          }
        })
    });
  }
  toCanvas(canvas) {
    return new Promise((resolve, reject) => {
      QRCode.toCanvas(canvas, this.value, {
        version: this.version,
        errorCorrectionLevel: this.errorCorrectionLevel,
        margin: this.margin,
        scale: this.scale,
        width: this.width,
        color: {
          dark: this.colorDark,
          light: this.colorLight
        }
      }, function (error) {
        if (error) {
          //console.error(error);
          reject(error);
        }
        else {
          //console.log('success!');
          resolve("success");
        }
      })
    });
  }
  renderElement(element) {
    for (let node of this.qrcElement.nativeElement.childNodes) {
      this.renderer.removeChild(this.qrcElement.nativeElement, node);
    }
    this.renderer.appendChild(this.qrcElement.nativeElement, element);
  }
  createQRCode() {
    if (!this.value) {
      return;
    };

    let element: Element;
    //console.log("QR Encoding " + this.value);

    switch (this.elementType) {

      case 'canvas':
        element = this.renderer.createElement('canvas');
        this.toCanvas(element).then((v) => {
          //console.log(v);
          this.renderElement(element);
        }).catch((e) => {
          console.error(e);
        });
        break;
      case 'url':
      case 'img':
      default:
        element = this.renderer.createElement('img');
        this.toDataURL().then((v: string) => {
          //console.log(v);
          element.setAttribute("src", v);
          element.setAttribute("alt", this.alt);
          this.renderElement(element);
        }).catch((e) => {
          console.error(e);
        })

    }


  }

}
